compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/ConversionUtils.kt

```diff
@@ -279,6 +296,64 @@ internal tailrec fun FirCallableSymbol<*>.deepestMatchingOverriddenSymbol(root:
     return overriddenSymbol.deepestMatchingOverriddenSymbol(this)
 }
 
+internal fun IrClass.findMatchingOverriddenSymbolsFromSupertypes(
+    irBuiltIns: IrBuiltIns,
+    target: IrDeclaration,
+    result: MutableList<IrSymbol> = mutableListOf()
+): List<IrSymbol> {
+    for (superType in superTypes) {
+        val superTypeClass = superType.classOrNull
+        if (superTypeClass is IrClassSymbolImpl) {
+            superTypeClass.owner.findMatchingOverriddenSymbolsFromThisAndSupertypes(irBuiltIns, target, result)
+        }
+    }
+    return result
+}
+
+private fun IrClass.findMatchingOverriddenSymbolsFromThisAndSupertypes(
+    irBuiltIns: IrBuiltIns,
+    target: IrDeclaration,
+    result: MutableList<IrSymbol>
+): List<IrSymbol> {
+    for (declaration in declarations) {
+        when {
+            declaration is IrFunction && target is IrFunction ->
+                if (declaration !is IrConstructor &&
+                    !declaration.isFakeOverride &&
+                    declaration.descriptor.modality != Modality.FINAL &&
+                    !Visibilities.isPrivate(declaration.visibility) &&
+                    isOverriding(irBuiltIns, target, declaration)
+                ) {
+                    result.add(declaration.symbol)
+                }
+            // TODO: property lookup to set overriddenSymbols for IrProperty (and accessors)?
+        }
+    }
+    // Stop traversing upwards if we find matching overridden symbols at this level.
+    if (result.isNotEmpty()) {
+        return result
+    }
+    return findMatchingOverriddenSymbolsFromSupertypes(irBuiltIns, target, result)
+}
+
+fun isOverriding(
+    irBuiltIns: IrBuiltIns,
+    target: IrFunction,
+    superCandidate: IrFunction
+): Boolean =
+    target.name == superCandidate.name &&
+            // Not checking the receiver type since we are traversing super type hierarchy, i.e., the expected subtype relation between
```

Comment:
It's not true IMO, extension receivers can have different types

Reply:
```
open class Base {
    fun String.foo() {}
}

class Derived : Base() {
    fun Int.foo() {}
}
```

Reply:
Oh, totally missed extension receiver; thank you for the fix!

But, my comment was mainly about the dispatch receiver type, so I believe it still holds (i.e., we still don't check the dispatch receiver type since their subtype relations are obvious).